elasticsearch/src/main/java/org/hibernate/search/elasticsearch/processor/impl/ElasticsearchWorkProcessor.java

```diff
@@ -87,313 +121,157 @@ public void close() {
+	 * @return The result of the given work.
 	 */
-	public void executeAsync(ElasticsearchWork<?> work) {
-		asyncProcessor.submit( Collections.singleton( work ) );
+	public <T> CompletableFuture<T> executeAsyncUnsafe(ElasticsearchWork<T> work) {
+		return start( work, parallelWorkExecutionContext );
 	}
 
 	/**
-	 * Execute a set of works asynchronously.
+	 * Return an orchestrator for non-stream background works.
+	 * <p>
+	 * Works submitted in the same changeset will be executed in the given order.
+	 * Relative execution order between changesets is undefined.
 	 * <p>
-	 * Works submitted in the same list will be executed in the given order.
+	 * Works submitted to this orchestrator
+	 * will only be bulked with subsequent works (possibly of a different changeset).
 	 * <p>
 	 * If any work throws an exception, this exception will be passed
-	 * to the error handler with an {@link ErrorContext} spanning at least the given works,
+	 * to the error handler with an {@link ErrorContext} spanning exactly the given works,
 	 * and the remaining works will not be executed.
 	 *
-	 * @param works The works to be executed.
+	 * @return the orchestrator for synchronous, non-stream background works.
 	 */
-	public void executeAsync(List<ElasticsearchWork<?>> works) {
-		asyncProcessor.submit( works );
-	}
+	public BarrierElasticsearchWorkOrchestrator createNonStreamOrchestrator(String indexName, boolean sync, boolean refreshAfterWrite) {
+		/*
+		 * Since works are applied in order, refreshing the index after changesets
+		 * is actually an option, and if enabled we use refreshing execution contexts.
+		 * In order to reduce the cost of those refreshes, we also try to batch together
+		 * refreshes for works bulked in the same bulk API call. Non-bulked works will have
+		 * their refresh executed at the end of each changeset.
+		 */
+		Supplier<FlushableElasticsearchWorkExecutionContext> contextSupplier;
+		boolean refreshInBulkApiCall;
+		if ( refreshAfterWrite ) {
+			contextSupplier = this::createRefreshingWorkExecutionContext;
+			refreshInBulkApiCall = true;
+		}
+		else {
+			contextSupplier = this::createIndexMonitorBufferingWorkExecutionContext;
+			refreshInBulkApiCall = false;
+		}
 
-	/**
-	 * Blocks until the queue of requests scheduled for asynchronous processing has been fully processed.
-	 * N.B. if more work is added to the queue in the meantime, this might delay the wait.
-	 */
-	public void awaitAsyncProcessingCompletion() {
-		asyncProcessor.awaitCompletion();
+		/*
+		 * The non-stream orchestrator requires a strict execution ordering,
+		 * because subsequent changesets may be inter-dependent (an addition in one changeset
+		 * followed by a deletion in the next, for instance).
+		 * To make sure we apply works in order, we use serial orchestrators.
+		 */
+		FlushableElasticsearchWorkOrchestrator delegate =
+				createSerialOrchestrator( contextSupplier, NON_STREAM_MIN_BULK_SIZE, refreshInBulkApiCall );
+
+		if ( sync ) {
```

Comment:
Turns out it is. There was other differences before, but I removed them along the commits.
The thing is, the orchestrators themselves work the same way whether we're in synchronous mode or not. It's just the index manager that will wait for submitted works to finish in synchronous mode, and won't wait in asynchronous mode.
I'll add a fixup commit.